examples/echo.rs

```diff
@@ -56,7 +48,7 @@ fn echo(req: Request<Body>) -> BoxFut {
                 response
             });
 
-            return Box::new(reversed);
+            return Ok(reversed);
```

Comment:
how to deal with `concat2` above, it also seems to be a function that is missing

`no method named `concat2` found for type `hyper::Body` in the current scope`

Reply:
It looks like that was renamed to `TryStreamExt::try_concat`.

Reply:
hmmm.. I tried to go with it, but not really sure how :P 
Got some errors like this with my last commit c04d92f
```
error[E0277]: `(dyn hyper::rt::Stream<Item = std::result::Result<hyper::Chunk, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)` cannot be shared between threads safely
  --> examples/echo.rs:78:10
   |
78 |         .serve(service);
   |          ^^^^^ `(dyn hyper::rt::Stream<Item = std::result::Result<hyper::Chunk, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)` cannot be shared between threads safely
   |
```

Reply:
Okay, I checked out the branch, and finally understand the giant error messages. They're very confusing, but ultimately, since there was an `await` inside the `match req.uri()`, the `await` was forcing one of the state machine steps to own a borrow of `&Request`. Because of that, it was noticing the `&Request` is not `Sync`, thanks to the wrapped stream variant inside `Body`.

This will be very easy for users to bump into, and so probably I'll need to fix `Body` to require `Sync`...

Reply:
For reference, here's the issue in the rust-lang repo: https://github.com/rust-lang/rust/issues/57017

Reply:
Ok, I've merged to master a fix for `Body` that should remove that particular issue.

Reply:
Well that sounds complicated. Not sure how much I could really understand but it's nice that you helping to solve it out :)

I removed onto master, now I seem to have this issue:
```
error[E0599]: no method named `into_future` found for type `tokio_sync::mpsc::bounded::Receiver<_>` in the current scope
   --> src/client/connect/dns.rs:134:14
    |
134 |             .into_future()
    |              ^^^^^^^^^^^
    |
    = note: the method `into_future` exists but the following trait bounds were not satisfied:
            `&mut tokio_sync::mpsc::bounded::Receiver<_> : futures_util::stream::StreamExt`
            `&tokio_sync::mpsc::bounded::Receiver<_> : futures_util::stream::StreamExt`
            `tokio_sync::mpsc::bounded::Receiver<_> : futures_util::stream::StreamExt`

error: aborting due to previous error
```

Reply:
You may need to run `cargo update` to get the latest changes to tokio.

Reply:
Hmmm... did that then I got this 😄 
```
error[E0277]: `dyn std::error::Error` cannot be sent between threads safely
  --> examples/echo.rs:67:1
   |
67 | #[tokio::main]
   | ^^^^^^^^^^^^^^ `dyn std::error::Error` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `dyn std::error::Error`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<dyn std::error::Error>`
   = note: required because it appears within the type `std::boxed::Box<dyn std::error::Error>`
   = note: required because it appears within the type `std::result::Result<(), std::boxed::Box<dyn std::error::Error>>`
```

Reply:
Oh cool, looks like I just need to add
`async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> ` on the main. Nice.

Reply:
looks like it is working now, let me tide up the PR a bit. Thanks for the help 😄